app/scripts/metamask-controller.js

```diff
@@ -5310,6 +5358,225 @@ export default class MetamaskController extends EventEmitter {
+   * @param {boolean} options.autoApprove - If the chain should be granted without prompting for user approval.
+   */
+  async requestPermittedChainsPermissionIncremental({
+    origin,
+    chainId,
+    autoApprove,
+  }) {
+    if (isSnapId(origin)) {
+      throw new Error(
+        `cannot request permittedChains permission for snaps with origin "${origin}"`,
+      );
+    }
+
+    const caip25Caveat = this.permissionController.getCaveat(
+      origin,
+      Caip25EndowmentPermissionName,
+      Caip25CaveatType,
+    );
+
+    if (caip25Caveat.value.isMultichainOrigin) {
+      throw providerErrors.unauthorized(
+        `Cannot switch to or add permissions for chainId '${chainId}' because permissions were granted over the Multichain API.`,
+      );
+    }
+
+    if (!autoApprove) {
+      await this.requestApprovalPermittedChainsPermission(origin, chainId);
+    }
+
+    const caveatValueWithChainsAdded = addPermittedEthChainId(
+      caip25Caveat.value,
+      chainId,
+    );
+
+    const ethAccounts = getEthAccounts(caip25Caveat.value);
+    const caveatValueWithAccountsSynced = setEthAccounts(
+      caveatValueWithChainsAdded,
+      ethAccounts,
+    );
+
+    this.permissionController.updateCaveat(
+      origin,
+      Caip25EndowmentPermissionName,
+      Caip25CaveatType,
+      caveatValueWithAccountsSynced,
+    );
+  }
+
+  /**
+   * Requests CAIP-25 for permissions for the specified origin
+   * and replaces any existing CAIP-25 permission with a new one.
+   *
+   * @param {string} origin - The origin to request approval for.
+   * @param requestedPermissions - The legacy permissions to request approval for.
+   * @returns the granted CAIP-25 Permission.
+   */
+  async requestCaip25Permission(origin, requestedPermissions = {}) {
+    const permissions = pick(requestedPermissions, [
+      RestrictedMethods.eth_accounts,
+      PermissionNames.permittedChains,
+    ]);
+
+    if (!permissions[RestrictedMethods.eth_accounts]) {
+      permissions[RestrictedMethods.eth_accounts] = {};
+    }
+
+    if (!permissions[PermissionNames.permittedChains]) {
+      permissions[PermissionNames.permittedChains] = {};
+    }
+
+    if (isSnapId(origin)) {
+      delete permissions[PermissionNames.permittedChains];
+    }
+
+    const id = nanoid();
+    const legacyApproval =
+      await this.approvalController.addAndShowApprovalRequest({
+        id,
+        origin,
+        requestData: {
+          metadata: {
+            id,
+            origin,
+          },
+          permissions,
+        },
+        type: MethodNames.RequestPermissions,
+      });
+
+    const newCaveatValue = {
+      requiredScopes: {},
+      optionalScopes: {},
+      isMultichainOrigin: false,
+    };
+
+    const caveatValueWithChains = isSnapId(origin)
+      ? {
+          ...newCaveatValue,
+          optionalScopes: {
+            'wallet:eip155': {
+              accounts: [],
```

Comment:
Not sure I follow this 🤔

Reply:
This block is only responsible for creating the empty `wallet:eip155` scope. Accounts themselves are set in the `setEthAccounts()` call below this

Reply:
How come we aren't adding `wallet:eip155` for dapps here?

Reply:
Doing so would be unnecessary since Dapps will have some chain specific `eip155:x` scope already. Additionally it would also have no affect in the Multichain API because this EIP-1193 granted permission does not grant access to the Multichain API (where `wallet:eip155` being granted would have an effect)

Reply:
This seems to differ from the migration where we are giving all origins `wallet:eip155`. I was under the impression that we were granting dapps `wallet:eip155` to prepare for a future where that scope might include signing methods. I'm not following the reasoning for why this isn't the case here.